### DISCUSSION TITLE
Handling IDisposable.

### DISCUSSION THREAD

**Discussion by Gavin-Williams** [amerkoleci/Vortice.Windows#336] [Q&A]
I've just discovered that VS has a quick action 'Implement interface with Dispose pattern' which is good, because I'm never quite sure how to implement it. Not that I understand the quick action pattern, but I'll just follow it's recommendations. However, the following questions come up..

Is Vortice object disposal considered managed or unmanaged? I'm not sure about the distinction, because presumably Vortice objects can internally configure whether they are picked up by the GC and handle any native object disposal appropriately.

And as a consequence of that. Should I uncomment this...
```
        // // TODO: override finalizer only if 'Dispose(bool disposing)' has code to free unmanaged resources
        // ~Device3d()
        // {
        //     // Do not change this code. Put cleanup code in 'Dispose(bool disposing)' method
        //     Dispose(disposing: false);
        // }
```

**Comment by DemoBytom** [comment]
I was also wondering about this issue lately. Looking through the code in this repository I can't see dispose pattern implemented anywhere, like: https://github.com/amerkoleci/Vortice.Windows/blob/9a6e9c42587cbba69a1614909a2424312adfa05b/src/samples/HelloDirect3D12/ShaderIncludeHandler.cs#L88-L96

I usually do implement the `IDisposable` pattern in any non-sealed classes, and for now I've been putting any `Vortice.Windows`/DirectX disposables into the "managed" part of the pattern, and I've been leaving the destructor commented out/deleted, like that:
```csharp
    private void Dispose(bool disposing)
    {
        if (!_disposedValue)
        {
            if (disposing)
            {
                // TODO: dispose managed state (managed objects)
                [...]
                _swapChain?.Dispose(); //IDXGISwapChain4
                _swapChain = null;
               [...]
            }

            // TODO: free unmanaged resources (unmanaged objects) and override finalizer
            // TODO: set large fields to null
        }
    }

    // // TODO: override finalizer only if 'Dispose(bool disposing)' has code to free unmanaged resources
    // ~RenderingSurface()
    // {
    //     // Do not change this code. Put cleanup code in 'Dispose(bool disposing)' method
    //     Dispose(disposing: false);
    // }

    public void Dispose()
    {
        // Do not change this code. Put cleanup code in 'Dispose(bool disposing)' method
        Dispose(disposing: true);
        GC.SuppressFinalize(this);
    }
```

But I am honestly unsure if that's the proper way to do it.
